week3_ts/src/pages/Main.tsx

```diff
@@ -0,0 +1,58 @@
+import { handsomeGuys, HandsomeGuy } from "@/assets/images";
+import MainHeader from "@/component/MainHeader";
+import Tournament from "@/component/Tournament";
+import { useEffect, useRef, useState } from "react";
+import { useNavigate } from "react-router-dom";
+import styled from "styled-components";
+
+export default function Main() {
+    // 배열 랜덤으로 재정렬
+    const randomGameInfo: HandsomeGuy[] = handsomeGuys.sort(() => Math.random() - 0.5);
+
+    const [round, setRound] = useState<string>("16강");
+    const matchWinners = useRef<HandsomeGuy[]>([]);
+    const [fighterList, setFighterList] = useState<HandsomeGuy[]>(randomGameInfo);
+    const navigation = useNavigate();
+
+    const checkLists = (round: string) => {
+        setRound(round);
+        setFighterList(matchWinners.current);
+        matchWinners.current = [];
+    };
+
+    // 화면이 리렌더링 될 때 마다 참가자들 배열과 승리자들 배열 확인
+
+    useEffect(() => {
+        if (fighterList.length === 0 && matchWinners.current.length >= 8) checkLists("8강");
+        else if (fighterList.length === 0 && matchWinners.current.length >= 4) checkLists("4강");
+        else if (fighterList.length === 0 && matchWinners.current.length >= 2) checkLists("결승");
+        else if (fighterList.length === 0 && matchWinners.current.length === 1) {
+            console.log(matchWinners.current[0]);
+            navigation("/complete", {
+                // navigation에 값을 다중으로 넘겨주는 방법
+                state: { matchWinners },
+            });
+        }
+    });
```

Comment:
`useEffect` 를 사용할 떄는 엔간해서 두번째 인자를 설정해주는 것이 좋다고 알고 있는데,
따로 이유가 있을까요??
[공식문서](https://ko.reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects)

useEffect  내에서는 figtherList와 matchWinner만 활용하는 데에 비해,
두번째 인자를 설정하지 않는다면 해당 변수 외에 모든 것이 변할 때 (리렌더링 될 떄) 마다 콜백을 실행시키게 됩니다
무한 루프에 갇히게 될 가능성도 있고, 불필요한 렌더링이 지속적으로 일어나게 될 것 같아요!

```suggestion
    useEffect(() => {
        if (fighterList.length === 0 && matchWinners.current.length >= 8) checkLists("8강");
        else if (fighterList.length === 0 && matchWinners.current.length >= 4) checkLists("4강");
        else if (fighterList.length === 0 && matchWinners.current.length >= 2) checkLists("결승");
        else if (fighterList.length === 0 && matchWinners.current.length === 1) {
            console.log(matchWinners.current[0]);
            navigation("/complete", {
                // navigation에 값을 다중으로 넘겨주는 방법
                state: { matchWinners },
            });
        }
    }, [fighterList, matchWinner.current]);
```

Reply:
useEffect를 사용할 때 두번째 인자를 붙이는 걸 자꾸 잊어버리는 거 같아요 제대로 사용하겠심다 !.!